File: frontend/shared/schemas.ts
/** Canonical domain schemas — frozen in Phase 0 (TDS §8). */

import type {
  AssessmentStatus,
  AssessmentType,
  DecisionOutcome,
  FactType,
  RecommendationDisposition,
  ReferenceSource,
  RetrievalMode,
  RetrievalPath,
  RetrievalQuality,
  ReviewState,
  RiskLevel,
} from "./enums";

export interface Context {
  id: string;
  asset_id: string;
  category: string;
  payload: Record<string, unknown>;
  provider: string;
  valid_from: string;
  valid_until: string;
  confidence: number;
}

export interface DerivedFact {
  id: string;
  asset_id: string;
  fact_type: FactType | string;
  value: boolean | number | string;
  computed_at: string;
  source_context_ids: string[];
}

export interface RetrievedReference {
  source: ReferenceSource;
  id: string;
  retrieval_path: RetrievalPath;
  score: number | null;
  chunk_id: string | null;
  title?: string | null;
  snippet?: string | null;
  code?: string | null;
  triggered_by_fact?: string | null;
  /** Primary-source link, so a cited clause can be checked rather than trusted. */
  source_url?: string | null;
}

export interface ReasoningFactor {
  fact_type: string;
  headline: string;
  detail: string;
  evidence: RetrievedReference[];
  context_ids: string[];
}

export interface AreaOwner {
  worker_id: string;
  name: string;
  role: string;
  zone: string;
}

export interface Recommendation {
  id: string;
  text: string;
  rationale: string;
  disposition: RecommendationDisposition | null;
}

export interface RecommendationIn {
  text: string;
  rationale: string;
}

export interface ManualAssessmentIn {
  summary: string;
  risk_level: RiskLevel;
  recommendations: RecommendationIn[];
}

export interface AssessmentMetadata {
  provider: string;
  model: string;
  prompt_version: string;
  input_tokens: number;
  output_tokens: number;
  estimated_cost_usd: number;
  latency_ms: number;
  timestamp: string;
  retrieved_context_ids: string[];
  retrieved_evidence_ids: string[];
  retrieval_mode: RetrievalMode;
  retrieval_quality: RetrievalQuality;
  retrieval_score: number | null;
  embedding_model: string | null;
  confidence: number;
  assessment_version: number;
  reasoning_factors?: ReasoningFactor[];
}

export interface Assessment {
  id: string;
  review_id: string;
  assessment_type: AssessmentType;
  status: AssessmentStatus;
  risk_level: RiskLevel;
  summary: string;
  recommendations: Recommendation[];
  derived_fact_ids: string[];
  metadata: AssessmentMetadata | null;
  reasoning_factors?: ReasoningFactor[];
}

export interface Decision {
  id: string;
  review_id: string;
  assessment_id: string;
  decided_by: string;
  outcome: DecisionOutcome;
  recommendation_dispositions: Record<string, "accepted" | "rejected">;
  conditions: string | null;
  comments: string | null;
  submitted_at: string;
}

export interface Review {
  id: string;
  asset_id: string;
  state: ReviewState;
  owner_id: string;
  triggered_by: string;
  origin: "system" | "operator" | "supervisor";
  raised_by_worker_id: string | null;
  created_at: string;
}

export interface Asset {
  id: string;
  name: string;
  zone: string;
  plant_id: string;
  floor: "ground" | "first" | "second";
}

export interface Report {
  id: string;
  review_id: string;
  closure_event_seq: number;
  content: Record<string, unknown>;
  generated_at: string;
}

export interface Notification {
  id: string;
  review_id: string | null;
  event_type: string;
  summary: string;
  recipient_ids: string[];
  created_at: string;
}

/**
 * Shift handover — a custody transfer between panel operators.
 *
 * A handover is `draft` while the outgoing operator edits it, `issued` once
 * handed to the incoming operator, and `accepted` only after every item marked
 * `requires_ack` has been acknowledged or queried. `expired` marks one that was
 * superseded before it was accepted — its pending items are the gaps.
 */
export type HandoverState = "draft" | "issued" | "accepted" | "expired";

export type HandoverItemType =
  | "open_review"
  | "active_fact"
  | "open_task"
  | "decision_condition"
  | "note";

export type HandoverAckState = "pending" | "acknowledged" | "queried";

/** Whether the brief prose came from a model or the deterministic template. */
export type HandoverNarrationMode = "llm" | "deterministic";

export interface HandoverItem {
  id: string;
  item_type: HandoverItemType;
  position: number;
  review_id: string | null;
  asset_id: string | null;
  asset_name: string | null;
  task_id: string | null;
  title: string;
  detail: string | null;
  risk_level: string;
  hazard_dimensions: string[];
  requires_ack: boolean;
  ack_state: HandoverAckState;
  ack_note: string | null;
  acknowledged_by: string | null;
  acknowledged_by_name: string | null;
  acknowledged_at: string | null;
  source: "auto" | "manual";
}

export interface Handover {
  id: string;
  state: HandoverState;
  outgoing_actor_id: string;
  outgoing_actor_name: string;
  incoming_actor_id: string;
  incoming_actor_name: string;
  window_start: string;
  window_end: string;
  brief: string | null;
  narration_mode: HandoverNarrationMode;
  issued_at: string | null;
  accepted_at: string | null;
  created_at: string;
  items: HandoverItem[];
  required_total: number;
  required_cleared: number;
  attention_asset_id: string | null;
  /** Which side of this handover the requester is on; supervisors are observers. */
  viewer_role: "outgoing" | "incoming" | "observer";
}

export interface HandoverGap {
  handover_id: string;
  item_id: string;
  asset_id: string | null;
  asset_name: string | null;
  title: string;
  risk_level: string;
  incoming_actor_name: string;
  issued_at: string | null;
  hours_outstanding: number;
}

export interface HandoverMetrics {
  handovers_total: number;
  handovers_accepted: number;
  required_items_total: number;
  required_items_cleared: number;
  coverage_pct: number;
  median_ack_minutes: number | null;
  unacknowledged_crossings: number;
}

File: backend/app/handover/schemas.py
"""Shift-handover request/response shapes."""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

HandoverState = Literal["draft", "issued", "accepted", "expired"]
HandoverItemType = Literal[
    "open_review", "active_fact", "open_task", "decision_condition", "note"
]
AckState = Literal["pending", "acknowledged", "queried"]
NarrationMode = Literal["llm", "deterministic", "fallback"]


class HandoverItemOut(BaseModel):
    id: UUID
    item_type: HandoverItemType
    position: int
    review_id: UUID | None = None
    asset_id: UUID | None = None
    asset_name: str | None = None
    task_id: UUID | None = None
    title: str
    detail: str | None = None
    risk_level: str
    hazard_dimensions: list[str] = Field(default_factory=list)
    requires_ack: bool
    ack_state: AckState
    ack_note: str | None = None
    acknowledged_by: UUID | None = None
    acknowledged_by_name: str | None = None
    acknowledged_at: datetime | None = None
    source: Literal["auto", "manual"]


class HandoverOut(BaseModel):
    id: UUID
    state: HandoverState
    outgoing_actor_id: UUID
    outgoing_actor_name: str
    incoming_actor_id: UUID
    incoming_actor_name: str
    window_start: datetime
    window_end: datetime
    brief: str | None = None
    narration_mode: NarrationMode
    issued_at: datetime | None = None
    accepted_at: datetime | None = None
    created_at: datetime
    items: list[HandoverItemOut] = Field(default_factory=list)
    required_total: int = 0
    required_cleared: int = 0
    #: The asset the incoming operator should look at first — highest-risk
    #: unacknowledged item, else the highest-risk item.
    attention_asset_id: UUID | None = None
    #: Which side of this handover the requesting actor is on. `observer` covers
    #: the supervisor, who can read a handover but is never a party to one.
    viewer_role: Literal["outgoing", "incoming", "observer"] = "observer"


class HandoverDraftIn(BaseModel):
    incoming_actor_id: UUID
    window_hours: int = Field(default=12, ge=1, le=72)


class HandoverNoteIn(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    detail: str | None = None
    requires_ack: bool = True


class HandoverAckIn(BaseModel):
    ack_state: Literal["acknowledged", "queried"] = "acknowledged"
    note: str | None = None


class HandoverGapOut(BaseModel):
    """A high-risk item that crossed a boundary without being acknowledged."""

    handover_id: UUID
    item_id: UUID
    asset_id: UUID | None = None
    asset_name: str | None = None
    title: str
    risk_level: str
    incoming_actor_name: str
    issued_at: datetime | None = None
    hours_outstanding: float


class HandoverMetricsOut(BaseModel):
    handovers_total: int
    handovers_accepted: int
    required_items_total: int
    required_items_cleared: int
    #: Share of high-risk carried items the incoming operator actually
    #: acknowledged. This is an operational measure of the handover process; it
    #: is deliberately not a detector input (see eval/detectors.py).
    coverage_pct: float
    median_ack_minutes: float | None = None
    unacknowledged_crossings: int = 0
